### ISSUE TITLE
Calling IQueryable-producing methods inside expression tree throws exception when outer is enumerated: "A second operation started on this context before a previous operation completed. Any instance members are not guaranteed to be thread safe."

### ISSUE THREAD

**Issue by jeremyZX** [aspnet/EntityFramework#7783]
When creating a query, if I call a method that returns an IQueryable inside an expression, when the query is enumerated, an exception is thrown, as though both the outer Queryable and inner Queryable are trying to enumerate simultaneously.

I can work around this by introducing a variable that represents the inner IQueryable, and reference that from within the Expression.  In prior versions of EntityFramework, though, up through at least EF Core RC1, the direct method call approach used to work without a hitch.

```
Exception message:

System.InvalidOperationException: A second operation started on this context before a previous operation completed. Any instance members are not guaranteed to be thread safe.

Stack trace:
Microsoft.EntityFrameworkCore.Internal.ConcurrencyDetector.EnterCriticalSection()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__15`2.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at TestApp.Services.BloggingService.GetBlogsWithPosts() in c:\users\jeremy\documents\visual studio 2015\Projects\TestApp\src\TestApp\Services\Services.cs
```

### Simple Repro Case

Using samples from https://docs.microsoft.com/en-us/ef/core/get-started/

```c#
public class Repository<T> where T : class
{
  // where BloggingContext is from https://docs.microsoft.com/en-us/ef/core/get-started/aspnetcore/new-db
  private readonly BloggingContext _bloggingContext;

  public Repository(BloggingContext bloggingContext)
  {
    _bloggingContext = bloggingContext;
  }

  public IQueryable<T> Find()
  {
    return _bloggingContext.Set<T>().AsQueryable();
  }
}

public class BloggingService
{
  private Repository<Blog> _blogRepository;
  private Repository<Post> _postRepository;
  
  public BloggingService(/*...*/) { /*...*/ }

  public IList<Blog> GetBlogsWithPostsFailMethodSyntax()
  {
    return _blogRepository.Find()
      .Where(blog => _postRepository.Find().Any(post => post.BlogId == blog.BlogId))
      .ToList(); // fails
  }

  public IList<Blog> GetBlogsWithPostsFailQuerySyntax()
  {
    return (from blog in _blogRepository.Find()
           where _postRepository.Find().Any(post => post.BlogId == blog.BlogId)
           select blog).ToList(); // fails
  }

  public IList<Blog> GetBlogsWithPostsSuccess()
  {
    var postQuery = _postRepository.Find();
    return _blogRepository.Find()
      .Where(blog => postQuery.Any(post => post.BlogId == blog.BlogId))
      .ToList(); // succeeds
  }
}
```

### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer 1.1.0
Operating system: Windows 10
IDE: Visual Studio 2015

**Comment by ajcvickers** [comment]
@anpete Can you investigate whether this is something that should work, but is being detected as a concurrency issue, or something that just won't work.

**Comment by arisliang** [comment]
so when is the fix planning to release?

**Comment by ajcvickers** [comment]
@arisliang As the milestone says, it will be included in the 2.0 release, which will be out pretty soon now.

**Comment by irowbin** [comment]
@ajcvickers I am also having issue with this query
```javascript

var parent = _service.NavigationStore.AsQueryable().Where( p => p.ParentId == null && p.ExtensionName.Equals( exn,StringComparison.OrdinalIgnoreCase ));
var child = _service.NavigationStore.AsQueryable().Where( p => p.ParentId > 0 && p.ExtensionName.Equals(exn, StringComparison.OrdinalIgnoreCase));

// joining with same table here
var data =  parent.GroupJoin( child, p => p.Id, c => c.ParentId,
    ( p, g ) => new { Parent = p, Children = g  } ).ToList();
```

Exception message **`A second operation started on this context before a previous operation completed. Any instance members are not guaranteed to be thread safe.`**

Am i doing wrong or its `entity framework` itself?
